src/clj/rems/workflow/dynamic.clj

```diff
@@ -35,12 +28,155 @@
+    ::return
+    ::save-draft
+    ::submit})
+(def EventTypes
+  #{#_:event/license-accepted
+    #_:event/license-required
+    :event/approved
+    :event/closed
+    :event/comment-requested
+    :event/commented
+    :event/decided
+    :event/decision-requested
+    :event/draft-saved
+    :event/member-added
+    :event/rejected
+    :event/returned
+    :event/submitted})
+
+;; TODO: namespaced keys e.g. :event/type, :event/time, :event/actor, :application/id
+;; TODO: add version number to events
+(s/defschema EventBase
+  {:event (apply s/enum EventTypes)
+   :application-id s/Int
+   :actor UserId
+   :time DateTime})
 
+(s/defschema ApprovedEvent
+  (assoc EventBase
+         :event (s/eq :event/approved)
+         :comment s/Str))
+(s/defschema ClosedEvent
+  (assoc EventBase
+         :event (s/eq :event/closed)
+         :comment s/Str))
+(s/defschema CommentedEvent
+  (assoc EventBase
+         :event (s/eq :event/commented)
+         :comment s/Str))
+(s/defschema CommentRequestedEvent
+  (assoc EventBase
+         :event (s/eq :event/comment-requested)
+         :commenters [s/Str]
+         :comment s/Str))
+(s/defschema DecidedEvent
+  (assoc EventBase
+         :event (s/eq :event/decided)
+         :decision (s/enum :approved :rejected)
+         :comment s/Str))
+(s/defschema DecisionRequestedEvent
+  (assoc EventBase
+         :event (s/eq :event/decision-requested)
+         :decider s/Str
+         :comment s/Str))
+(s/defschema DraftSavedEvent
+  (assoc EventBase
+         :event (s/eq :event/draft-saved)
+         :items {Long s/Str}
+         :licenses {Long s/Str}))
+(s/defschema MemberAddedEvent
+  (assoc EventBase
+         :event (s/eq :event/member-added)
+         :member s/Str))
+(s/defschema RejectedEvent
+  (assoc EventBase
+         :event (s/eq :event/rejected)
+         :comment s/Str))
+(s/defschema ReturnedEvent
+  (assoc EventBase
+         :event (s/eq :event/returned)
+         :comment s/Str))
+(s/defschema SubmittedEvent
+  (assoc EventBase
+         :event (s/eq :event/submitted)))
+
+(def event-schemas
+  [ApprovedEvent
+   ClosedEvent
+   CommentedEvent
+   CommentRequestedEvent
+   DecidedEvent
+   DecisionRequestedEvent
+   DraftSavedEvent
+   MemberAddedEvent
+   RejectedEvent
+   ReturnedEvent
+   SubmittedEvent])
 
+(s/defschema Event
+  (let [preds-and-schemas (->> event-schemas
+                               (map (fn [schema]
+                                      (let [event-type (:v (:event schema))]
+                                        (assert (keyword? event-type)
+                                                (str "couldn't get the event type from schema " schema))
+                                        [(fn [event]
+                                           (= event-type (:event event)))
+                                         schema])))
+                               (apply concat))]
+    (apply s/conditional (concat preds-and-schemas ['unknown-event-type]))))
```

Comment:
this seems pretty complex. isn't there some combinator in schema that we can use?